### DISCUSSION TITLE
Important Debugging Steps

### DISCUSSION THREAD

**Discussion by thenetworkgrinch** [BroncBotz3481/YAGSL-Example#47] [General]
When you are tuning your absolute encoder offsets **FACE THE BEVELS THE SAME WAY**.

If your odometry is always a bit off please check you wheel diameters, the grip tape might be worn and smaller than before.

**DISABLE HEADING CORRECTION IF CONTROLLING ROBOT HEADING VIA VELOCITY INSTEAD OF DESIRED ANGLE**

Make sure your absolute encoder is incrementing in the right direction! If your wheels never get to the right alignment they may need to be inverted!

If your wheels are constantly re-adjusting double check your angle gear ratio's.

**Comment by thenetworkgrinch** [comment]
Please make sure the magnets in your magnetic encoders (if you are using them) are glue'd into the shaft and have a good reading (green light on CANCoders)

**Reply by tfield** [reply]
One of ours came unglued and it was quite a surprise... but easily identified - wheels that were perfectly aligned one day were not aligned the next.  New measurement confirmed the issue and glue fixed the issue.

**Comment by thenetworkgrinch** [comment]
If you are using a NavX please recalibrate in the orientation you are using it in. ONCE and make sure the robot doesn't move!

**Reply by hungrypenguin77** [reply]
As I understand it, NavXs are inverted, with CW+ instead of CCW+.  As a result do we need to additionally invert the gyro?  Or does selecting the NavX already accomplish this?

**Reply by thenetworkgrinch** [reply]
> As I understand it, NavXs are inverted, with CW+ instead of CCW+. 

It should accomplish this but does not currently. Honestly I was unaware that the NavX's are CW+ not CCW+. After the season we should make that change.

> As a result do we need to additionally invert the gyro? Or does selecting the NavX already accomplish this?

As for now you should invert the gyro selecting the NavX **does not** already accomplish this.

**Comment by thenetworkgrinch** [comment]
Your gyroscope should be going Counter Clockwise Positive.

**Comment by thenetworkgrinch** [comment]
Your module locations are in inches and `y` = `left`, `x` = `front`

**Comment by thenetworkgrinch** [comment]
If your wheels spin out while running try increasing your amperage as a temporary fix, but there may be something wrong with a swerve module.

**Comment by thenetworkgrinch** [comment]
If the drive train slows down while at a rotation and strafing increase your current limits.

**Comment by thenetworkgrinch** [comment]
Tuning PID's..
![image](https://user-images.githubusercontent.com/10247070/227077214-89506e12-a897-46d2-922a-e2c0ddf3e242.png)

**Reply by thenetworkgrinch** [reply]
https://youtu.be/qKy98Cbcltw

**Reply by thenetworkgrinch** [reply]
https://file.tavsys.net/control/controls-engineering-in-frc.pdf

Section 2.6
![Screen_Shot_2023-03-26_at_12.43.37_PM.png](https://user-images.githubusercontent.com/10247070/227794595-bb566683-1fa4-4bf3-82b5-f3761a85d352.png)

**Comment by thenetworkgrinch** [comment]
Ensure the robot heading is correct by rotating the robot and looking at the gyro on Shuffleboard. The angle should increase while moving counter clockwise.

**Comment by thenetworkgrinch** [comment]
Advantage Scope (https://github.com/Mechanical-Advantage/AdvantageScope/releases)
or
frc-web-components (https://github.com/frc-web-components/app/tags)

to see what the swerve drive is commanded to do vs what it actually does. Values are published to `SmartDashboard/swerve`.

**Comment by thenetworkgrinch** [comment]
If your robot is always drifting to one side check your weight distribution!

**Comment by thenetworkgrinch** [comment]
If your modules periodically re-adjust themselves to different angles check to make sure your absolute encoders are the correct ones corresponding to your angle motors.

**Comment by thenetworkgrinch** [comment]
x and y are measured from center of wheel to center of robot. 

Robot odometry will not be accurate if this is incorrect especially during auto.

**Comment by thenetworkgrinch** [comment]
If this happens, you should
1. Ensure the correct ID's for motors and absolute encoders.
2. Ensure your maximum speed isn't too high.

https://user-images.githubusercontent.com/10247070/230080145-a2f15b3e-abef-4b1f-9d70-3e68098ad141.mov

**Comment by thenetworkgrinch** [comment]
If your swerve drive is inverting when changing orientation (front and back change) try inverting your drive motors. Like this.

https://user-images.githubusercontent.com/10247070/230226354-c7c5d0d2-3ac5-4f2f-96e8-f0d571d043ec.MOV



You may or may not have to invert your gyroscope following this.

**Comment by thenetworkgrinch** [comment]
If you need to invert your gyro but doing so messes up your driving try setting P to negative in controllerproperties.json.

**Comment by thenetworkgrinch** [comment]
If you need to invert your motors but doing so messes up your driving try setting P to negative in pidfproperties.json.

**Comment by thenetworkgrinch** [comment]
GREASE YOUR GEARS!!!

At least every day or once every 4 matches.

https://www.chiefdelphi.com/t/frc-95-the-grasshoppers-2023-build-thread/419845/137

**Comment by thenetworkgrinch** [comment]
Robot cannot turn 180 with path planner.

Try inverting your drive motors.

https://user-images.githubusercontent.com/10247070/232489640-cc4f5d30-2c77-43fe-97db-ac5aab63f4d5.mov

https://user-images.githubusercontent.com/10247070/232489666-f8811a54-b862-4fd9-8969-d682a0b429ec.mov

**Comment by thenetworkgrinch** [comment]
If pth planner is consistently traveling a offset, check wheel diameters

**Comment by MongooseTNM** [comment]
Hello, member of 2811 here. We got Yagsl running with navx, neos and srx mag encoders on a swerve X chassis. For some reason, the right stick does nothing and doesn't make the robot turn. I have tried to switch the axis it uses in robotcontainer but that didn't improve anything. Though when pressing B on the logitech f310 (button 1 if memory serves) it rotates sometimes.
Also the srx mag encoders only work when set to cancoders in the json file and they are connected to the dio pins on the roborio via and adapter board. Not really an issue but an interesting "bug".
Finally, the wheels turn 180 degrees to drive the robot backward instead of just driving the motor backward. You can see this when the robot changes direction and oscillates a little.
The issue with drive being inverted in the video was fixed by inverting both drive and turning in the json files. 

Any help is very appreciated.

https://github.com/BroncBotz3481/YAGSL-Example/assets/66395216/b8a3179b-6a40-40c5-93e4-cf4da89a78a3

**Reply by thenetworkgrinch** [reply]
I have not had any access to a Swerve X chassis or TalonSRX mag encoders on this setup. What does it look like in frc-web-components or advantagescope?

Your interesting bug is a direct correlation to the fall-back on the relative encoders. Essentially you may be running with no absolute encoders because the SRX Mag Encoder wrapper doesnt work or isnt configured correctly.

The 180 issue has been identified before and honestly i need to rewrite the handling and make a patched Rotation2d object which retains the original angle instead of a scoped one causing unpredictable 180 turns due to no context of 180 or 0 being closest

**Reply by MongooseTNM** [reply]
the srx mag encoder does work though because the code complains that the magnetic field is not optimal. I have not tried to use advantage scope but in the sim the robot turns correctly.

**Reply by thenetworkgrinch** [reply]
That error was generated by YAGSL because i expected CTRE to crash if the object didnt exist, fundamentally when you set your encoder type it goes through this switch statement.

https://github.com/BroncBotz3481/YAGSL-Example/blob/main/src/main/java/swervelib/parser/json/DeviceJson.java#L55-L61

Then it will create this object.

https://github.com/BroncBotz3481/YAGSL-Example/blob/main/src/main/java/swervelib/encoders/AnalogAbsoluteEncoderSwerve.java#L22-L40

It does expect the sensor to be attached to an analog input, which i may need to document better.

**Reply by MongooseTNM** [reply]
ok makes sense. any idea why the robot doesn't turn when given input to?

**Reply by thenetworkgrinch** [reply]
Btw the error is generated here, honestly not sure why you would see any of the other messages.


https://github.com/BroncBotz3481/YAGSL-Example/blob/main/src/main/java/swervelib/encoders/CANCoderSwerve.java#L90-L94


As for your code i would need to see it. But since the sim works i am not inclined to believe its your fault

**Reply by thenetworkgrinch** [reply]
Have you tripple checked your axis and button mappings or implemented it with a project not based off of YAGSL-Example's RobotContainer and commands?

**Reply by MongooseTNM** [reply]
I have triple-checked and tried switching the axis it uses and nothing helped. I have not tried using it with a project not based off the yagsl example robot container though

**Reply by thenetworkgrinch** [reply]
Check out what the widget in frc-web-components looks like and send it if you can!

**Reply by MongooseTNM** [reply]
which widget are you referencing?

**Comment by thenetworkgrinch** [comment]
Ensure your motor ID's are bellow 40.

**Comment by thenetworkgrinch** [comment]
Extreme network utilization could cause issues which disable motors temporarily. Use `SwerveTelemetry.verbosity` to address these issues!
